bigframes/core/rewrite.py

```diff
@@ -385,46 +385,72 @@ def common_selection_root(
     return None
 
 
+def pullup_limit_from_slice(
+    root: nodes.BigFrameNode,
+) -> Tuple[nodes.BigFrameNode, Optional[int]]:
+    """
+    This is a BQ-sql specific optimization that can be helpful as ORDER BY LIMIT is more efficient than WHERE + ROW_NUMBER().
+
+    Only use this if writing to an unclustered table. Clustering is not compatible with ORDER BY.
+    """
+    if isinstance(root, nodes.SliceNode):
+        new_child, limit = pullup_limit_from_slice(root.child)
+        # head case
+        # More cases could be handled, but this is by far the most important, as it is used by df.head(), df[:N]
+        if root.is_limit:
+            assert not root.start
+            assert root.step == 1
+            assert root.stop is not None
+            limit = root.stop
+            new_root, prior_limit = pullup_limit_from_slice(root.child)
+            if (prior_limit is not None) and (prior_limit < limit):
+                limit = prior_limit
+            return new_root, limit
+    elif (
+        isinstance(root, (nodes.SelectionNode, nodes.ProjectionNode))
+        and root.row_preserving
+    ):
+        new_child, limit = pullup_limit_from_slice(root.child)
+        if limit is not None:
+            return root.transform_children(lambda _: new_child), limit
+    # Most ops don't support pulling up slice, like filter, agg, join, etc.
+    return root, None
+
+
 def replace_slice_ops(root: nodes.BigFrameNode) -> nodes.BigFrameNode:
     # TODO: we want to pull up some slices into limit op if near root.
     if isinstance(root, nodes.SliceNode):
         root = root.transform_children(replace_slice_ops)
-        return convert_slice_to_filter(cast(nodes.SliceNode, root))
+        return rewrite_slice(cast(nodes.SliceNode, root))
     else:
         return root.transform_children(replace_slice_ops)
 
 
-def get_simplified_slice(node: nodes.SliceNode):
-    """Attempts to simplify the slice."""
-    row_count = traversals.row_count(node)
-    start, stop, step = node.start, node.stop, node.step
+def rewrite_slice(node: nodes.SliceNode):
```

Comment:
`rewrite_slice` seems like a good candidate for some unit tests if we don't have some already.

Reply:
ok, added some unit tests. Bit rough writing tree tests right now. Might need to setup a little framework if we want to do lots of these. Local engine could also be a way to ensure that tree rewrites do not change result value.